pycqed/instrument_drivers/physical_instruments/QuTech_AWG_Module.py

```diff
@@ -0,0 +1,371 @@
+        return self.ask('wlist:waveform:type? "%s"' % name)
+
+    def getWaveformLength(self, name):
+        '''
+        Args:
+            name (string):  waveform name excluding double quotes, e.g.
+            '*Sine100'
+        '''
+        return self.ask_int('wlist:waveform:length? "%s"' % name)
+
+    def newWaveformReal(self, name, len):
+        '''
+        Args:
+            name (string):  waveform name excluding double quotes, e.g.
+            '*Sine100'
+
+        NB: seems to do nothing (on Tek5014) if waveform already exists
+        '''
+        self.write('wlist:waveform:new "%s",%d,real' % (name, len))
+
+    def getWaveformData(self, name):
+        '''
+        Args:
+            name (string):  waveform name excluding double quotes, e.g.
+            '*Sine100'
+
+        Returns:
+            tuple containing lists: (waveform, marker1, marker2)
+
+        Compatibility:  5014, QWG
+        '''
+        self.write('wlist:waveform:data? "%s"' % name)
+        binBlock = self.binBlockRead()
+        # extract waveform and markers
+        waveformLen = len(binBlock)/5   # 5 bytes per record
+        waveform = []
+        marker1 = []
+        marker2 = []
+        for k in range(waveformLen):
+            (waveform[k], markers) = struct.unpack(binBlock, '<fB')
+            marker1[k] = markers & 0x01
+            marker2[k] = markers >> 1 & 0x01
+
+        return (waveform, marker1, marker2)
+
+    def sendWaveformDataReal(self, name, waveform, marker1, marker2):
+        """
+        send waveform and markers directly to AWG memory, i.e. not to a file
+        on the AWG disk.
+        NB: uses real data normalized to the range from -1 to 1 (independent
+        of number of DAC bits of AWG)
+
+        Args:
+            name (string): waveform name excluding double quotes, e.g. 'test'.
+            Must already exits in AWG
+
+            waveform (float[numpoints]): vector defining the waveform,
+            normalized between -1.0 and 1.0
+
+            marker1 (int[numpoints]): vector of 0 and 1 defining the first
+            marker
+
+            marker2 (int[numpoints]): vector of 0 and 1 defining the second
+            marker
+
+        Compatibility:  5014, QWG
+
+        Based on:
+            Tektronix_AWG5014.py::send_waveform, which sends data to an AWG
+            _file_, not a memory waveform
+            'awg_transferRealDataWithMarkers', Author = Stefano Poletto,
+            Compatibility = Tektronix AWG5014, AWG7102
+        """
+
+        # parameter handling
+        if len(marker1) == 0 and len(marker2) == 0:  # no marker data
+            m = np.zeros(len(waveform))
+        else:
+            if (not((len(waveform) == len(marker1))
+                    and ((len(marker1) == len(marker2))))):
+                raise UserWarning('length mismatch between markers/waveform')
+            # prepare markers
+            m = marker1 + np.multiply(marker2, 2)
+            m = int(np.round(m[i], 0))
```

Comment:
@wvlothuizen 
Can you explain what this line is supposed to do? 
There are some strange things, the variable `i` in  `m[i]` is not defined anywhere, 
It is unclear to me why m = marker1 + 2*marker2 (I assume that is what that line is supposed to do). 
Converting the array to an int will also yield weird behaviour. 

Stuff I'm doing now uses no markers but only waveforms so we're still fine but I think we need to discuss this hard linking of marker and waveform channels but that is a separate discussion.